Enumerate PhysicalDrives instead of LogicalVolumes on Windows

GregMefford opened the pull request:
In Windows, a LogicalVolume is a partition or collection of disks (e.g. a RAID group) that has a drive letter. I think what fwup would want to write to is the raw disk itself, which is a PhysicalDrive.

These are conveniently numbered from 1 to N as they are enumerated by the system. Unfortunately, there's not a reliable API to tell you which ones exist. They can disappear (e.g. by unplugging a USB device), leaving a gap in the sequence, as the higher-numbered PhysicalDrives are not re-numbered.

This implementation simply walks across the numbered drives until it finds 5 in a row that are invalid (do not exist). This eliminates the need to assume there won't be more than N drives and checking that many, and would only cause trouble if someone plugged in more than 5 card readers and then un-plugged all but the last one.

fhunleth commented:
The idea of using PhysicalDrives sounds good to me. I'm unsure about the 5 in a row part. If you unplug, plug, unplug, etc. over 5 times, can you get a high numbered PhysicalDrive that would fail? Does it take long to try all possible PhysicalDrives? I.e. is it possible to test every one fast enough that the user wouldn't notice?

GregMefford commented:
If you unplug and re-plug, it uses the same number again, so I don't think it's a concern. I don'believe there is a limit on the number of drives (at least not documented), so it seems silly to test 255 or 65535 drives. I'm certain it is possible to have a very high number of PhysicalDrives as opposed to LogicalDrives, because LogicalDrives have to have a letter assigned (of the 26 choices), whereas a JBOD disk shelf can hold dozens of drives that all appear separately to Windows.

fhunleth commented:
My comment is then, why 5? If we get a bug report that a PhysicalDrive isn't being detected, how do we know if it's not because of the gap of 5 test? It seems easier to debug if someone says PhysicalDrive256 isn't detected or whatever other limit we set. Plus, are we actually saving much time with this optimization? It may seem silly to poll 255 (or whatever number - 64K seems absurd, though) drives, but is it nanoseconds that we're saving?

fhunleth commented on src/mmc_windows.c:
```diff
@@ -78,63 +41,56 @@ static int mmc_query_volume_path(const WCHAR *volume_path, struct mmc_device *de
-    BOOL   Success              = FALSE;
-    WCHAR  VolumeName[MAX_PATH] = L"";
+    WCHAR drive_path[MAX_PATH] = L"";
+    HANDLE drive_handle;
```
Move these variables down their first use. I know that this isn't the typical Windows style for C code, but at least for the rest of fwup, I take advantage of C99's support for this. I prefer the style that reduces variable scope and the distance between declaration and use as much as is reasonable.

fhunleth commented on src/mmc_windows.c:
```diff
@@ -78,63 +41,56 @@ static int mmc_query_volume_path(const WCHAR *volume_path, struct mmc_device *de
+    // assuming there won't be any more past that point.
     int device_count = 0;
+    int consecutive_invalid_drives = 0;
+    for(int i=0; consecutive_invalid_drives < 5 && device_count < max_devices ; i++) {
```
Fix the spacing here: `for (int i = 0; consecutive_invalid_drives < 5 && device_count < max_devices; i++) {`

fhunleth commented on src/mmc_windows.c:
```diff
@@ -78,63 +41,56 @@ static int mmc_query_volume_path(const WCHAR *volume_path, struct mmc_device *de
     int device_count = 0;
+    int consecutive_invalid_drives = 0;
+    for(int i=0; consecutive_invalid_drives < 5 && device_count < max_devices ; i++) {
+        wsprintf(drive_path, L"\\\\.\\PhysicalDrive%d", i);
```
I'm pretty sure that all of the wide character operations here can be removed since the paths are guaranteed to be ASCII. I think that all that you need to do is remove the `#define UNICODE` at the top and don't use wide character versions of all of the `W` functions.

GregMefford commented:
I tried just scanning PhysicalDrive0 - 255 instead of trying to be clever, and it's pretty much instantaneous either way, so I went with that instead. Thanks for suggesting that. If someone had a problem where the 300th drive on their laptop wasn't showing up as an option, at least the troubleshooting would be pretty straightforward. :stuck_out_tongue_winking_eye:

I also addressed your other comments except the Unicode one. Does fwup guarantee that file paths won't use Unicode? What if someone has a folder or file name that uses Unicode characters?

fhunleth commented:
fwup's main() function takes non-wide chars, so if they're Unicode, they'll be UTF8 and that's how any code in this file would get it. I actually don't know what will happen on Windows if someone passes a Unicode filename, though. However, the paths for the code in this PR are all of the form `\\.\PhysicalDrive%d`, so using the Windows ANSI API is fine and it will let you delete a bunch of code does the character conversions.

fhunleth commented:
BTW, I just realized that I made the assumption that writing directly to `\\.\PhysicalDrive%d` in `fwup` would work. Did you try this yet? I don't know if you've gotten to it, but is it easy to automatically unmount (or whatever Windows calls it) the filesystems on the drive so that Windows doesn't flush its cache on top of us?

GregMefford commented:
Yeah, I'm pretty sure fwup apply doesn't work as it is now. I've been reading about how you're supposed to write to physical drives and it seems that there's a special locking procedure and you have to write in whole blocks at a time. It'll be my next focus to figure out the details and find a way to integrate it at seamlessly as possible into the rest of fwup.

GregMefford commented:
I did notice that there's already some code to buffer writes into blocks for more efficienct writes, so hopefully I can leverage that.

fhunleth commented:
It depends on the block size. Currently, fwup guarantees chunks of 512 bytes at 512 byte offsets. Fwup will often write larger blocks, but you can't count on it. My guess is that this will be ok on everything but the 4K hard disks. It is fine with me if we don't support those.

GregMefford commented:
With these last few commits, it's successfully acting like it's writing to the SD card without throwing errors.

It definitely did _something_ because Windows shows a formerly-working standard-partitioned Nerves SD card as simply "RAW" after I manually applied the 1K firmware test firmware file to it. I still need to validate that it's actually writing the card properly.

GregMefford commented:
Wow! I just burned `blinky.fw` from Windows, put it in my Raspberry Pi, and it worked! :tada: :rocket: 

```
C:\temp>fwup.exe -a -d \\.\PhysicalDrive5 -i blinky.fw -t complete
100%
```

fhunleth commented on src/mmc_windows.c:
```diff
@@ -33,163 +33,162 @@ void mmc_finalize()
+    WCHAR volume_name[MAX_PATH] = L"";
+    HANDLE volume_iter = FindFirstVolume(volume_name, ARRAYSIZE(volume_name));
+
+    if (volume_iter == INVALID_HANDLE_VALUE) {
```
No brace for one line blocks. I mostly follow the Qt coding style in my C and C++ projects: http://wiki.qt.io/Qt_Coding_Style.

fhunleth commented on src/mmc_windows.c:
```diff
@@ -33,163 +33,162 @@ void mmc_finalize()
-        fwup_warnx("Cannot open '%s'", mmc_device);
-        return -1;
+        fwup_errx(EXIT_FAILURE, "Error dismounting '%S', (Error %lu)", volume_name, GetLastError());
+        return;
```
No return is necessary here since `fwup_errx` exits.

fhunleth commented on src/mmc_windows.c:
```diff
@@ -33,163 +33,162 @@ void mmc_finalize()
+ * @param volume_name the name of the volume as a Wide String
+ * @return the PhysicalDrive Number on success, 0 on failure
+ */
+unsigned int query_physical_extents(LPWSTR volume_name) {
```
Make `query_physical_extents` a static function since it's not used outside of this file.

fhunleth commented on src/mmc_windows.c:
```diff
@@ -33,163 +33,162 @@ void mmc_finalize()
+ * @brief Attempt to dismount the specified volume, exit with failure message if unsuccessful
+ * @param volume_name the name of the volume as a Wide String
+ */
+void dismount_volume(LPWSTR volume_name) {
```
Use static for `dismount_volume`. Also, is `dismount` the Windows terminology? fwup uses `unmount` everywhere else, and it would be nice to be consistent unless it would confuse Windows users.

GregMefford replied:
I was on the fence between `unmount` and `dismount` because the IOCTL command is called dismount but standard fwup terminology is unmount. I went with `dismount` simply because that's the Windows API that it calls, but I could really go either way.

fhunleth replied:
My preference would be to have everything user facing, like the error messages, use the word `unmount`. Using either word `unmount` or `dismount` for function names in this one file is fine.

GregMefford commented:
I fixed the items in your comments and also added an attempt to lock the LogicalVolume before unmounting it. I believe you're supposed to do this but it seemed like it was working without it, so I only print a warning if the locking step fails for some reason.

fhunleth commented:
I'm not able to test it yet, but it looks good. Thanks for addressing the comments.
